### ISSUE TITLE
State changes not picked up

### ISSUE THREAD

**Issue by mjmeintjes** [roman01la/citrus#23]
Not sure if I'm missing something obvious, but I am having trouble getting this to work.

When I dispatch 2 events on the same controller, one after the other, they seem to overwrite each other's state.

Looking at the code for Reconciler, it looks like this is because the state for the controller gets extracted when enqueue-ing, and because of this it does not take any changes in state into account when running each event fn.

Is this expected behaviour?

**Comment by gmp26** [comment]
Just tested by patching the example counter with:
```clj
[:button {:on-click #(doseq [i (range 10)] (citrus/dispatch! r :counter :inc))} "+10"]
```
and yes, the counter only increments by 1.

**Comment by gmp26** [comment]
OK, I'm now of the opinion this is the behaviour we should expect for the asynchronous case as forcing a wait on state handling before enqueuing another event would be equivalent to synchronous behaviour.

**Comment by roman01la** [comment]
This is definitely a bug. Async events are batched and executed asynchronously, but this should not change the behavior. I’ll look into this.

**Comment by mjmeintjes** [comment]
Here's what I'm using for now:
``
(r/dispatch! [this cname event args]
    (queue-effects!
      queue
      [cname #((get controllers cname) event args (get % cname))])

    (schedule-update!
      batched-updates
      scheduled?
      (fn []
        (let [effects (into [] @queue)]
          (clear-queue! queue)
          (let [[curr-state handlers :as result] (reduce
                                       (fn [[curr-state handlers] [cname ctrl]]
                                         (let [effects (ctrl curr-state)]
                                           [(assoc curr-state cname (:state effects))
                                            (concat handlers (->> (dissoc effects :state)

                                                                  (filter (comp seq second))))]
                                           ))
                                       [@state []]
                                       effects)]
            (reset! state curr-state)
            (m/doseq [[id effect] handlers]
              (when-let [handler (get effect-handlers id)]
                (handler this cname effect))))))))
``

**Comment by gmp26** [comment]
@roman01la - yes - much easier to think about that way! Code from @mjmeintjes fixes my test case.

**Comment by roman01la** [comment]
Just published a fix in `3.0.1`. @mjmeintjes thank you for that code sample
